proposals/nnnn-actors.md

```diff
@@ -744,3 +677,64 @@ Nearly all changes in actor isolation are breaking changes, because the actor is
+
+```swift
+class OddOddySync {
+  let evan: EvenEvanSync!
+
+  func isOdd(_ n: Int) -> Bool {
+    if n == 0 { return true }
+    return evan.isEven(num - 1)
+  }
+}
+
+class EvenEvanSync {
+  let oddy: OddOddySync!
+
+  func isEven(_ n: Int) -> Bool {
+    if n == 0 { return false }
+    return oddy.isOdd(num - 1)
+  }
+}
+```
+
+This code is depending on the two methods of these classes to effectively be "reentrant" within the same call stack, because one will call into the other (and vice-versa) as part of the computation. Now, take this example and make it asynchronous using actors:
+
+```swift
+actor class OddOddy {
+  let evan: EvenEvan!
+
+  func isOdd(_ n: Int) async -> Bool {
+    if n == 0 { return true }
+    return await evan.isEven(num - 1)
+  }
+}
+
+actor class EvenEvan {
+  let oddy: OddOddy!
+
+  func isEven(_ n: Int) async -> Bool {
+    if n == 0 { return false }
+    return await oddy.isOdd(num - 1)
+  }
+}
+```
+
+Under the current proposal, this code will deadlock, because a call from `EvanEvan.isEven` to `OddOddy.isOdd` will then depend on another call to `EvanEvan.isEven`, which cannot proceed until the original call completes. One would need to make these methods reentrant to eliminate the deadlock.
+
+With Swift embracing [Structured Concurrency](https://github.com/DougGregor/swift-evolution/blob/structured-concurrency/proposals/nnnn-structured-concurrency.md) as a core building block of its concurrency story, we may be able to do better than outright banning reentrancy. In Swift, every asynchronous operation is part of a `Task` which encapsulates the general computation taking place, and every asynchronous operation spawned from such task becomes a child task of the current task. Therefore, it is possible to know whether a given asynchronous call is part of the same task hierarchy, which is the rough equivalent to being in the same call stack in synchronous code.
+
+We could introduce a new kind of reentrancy, *task-chain reentrancy*, which allows reentrant calls on behalf of the given task or any of its children. This resolves both the deadlock we encountered in the `Waiter` and `Kitchen` example from the section on [deadlocks](#deadlocks-with-non-reentrant-actors) as well as the mutually-recursive `isEven` example above, while still preventing reentrancy from unrelated tasks. This reentrancy therefore mimics synchronous code more closely, eliminating many deadlocks without allow unrelated interleavings to break the high-level invariants of an actor.
+
+There are a few reasons why we are not currently comfortable including task-chain reentrancy in the proposal:
+* The task-based reentrancy approach doesn't seem to have been tried at scale. Orleans documents support for [reentrancy in a call chain](https://dotnet.github.io/orleans/docs/grains/reentrancy.html#reentrancy-within-a-call-chain), but the implementation is fairly limited.
```

Comment:
You could even add:


What's more, it actually is _not present_ in the current Orleans runtime, as it was confusing and disapointing in real world scenarios, as mentioned by the current project's lead developer Reuben Bond [here](https://twitter.com/reubenbond/status/1349725703634251779).

In case you'd like to shut down people from trying to invent reasons that it's awesome and "orleans has it". :)

Reply:
Good point, I'll include this note, because it's important that they tried it and it didn't work out for them. That doesn't necessarily mean it's wrong for Swift, but it's a concern!

Reply:
Yup, exactly. It's a bit too tempting to point at it and wave hands around "make it work, it's magic" otherwise :D